src/app/cli/src/coda.ml

```diff
@@ -9,6 +9,123 @@ open Signature_lib
 open Init
 module YJ = Yojson.Safe
 
+let retrieve_genesis_state dir_opt ~logger :
+    (Ledger.t lazy_t * Proof.t) Deferred.t =
+  let open Cache_dir in
+  let tar_filename = Cache_dir.genesis_dir_name ^ ".tar.gz" in
+  let s3_bucket_prefix =
+    "https://s3-us-west-2.amazonaws.com/snark-keys.o1test.net" ^/ tar_filename
+  in
+  let extract dir =
+    match%map
+      Monitor.try_with_or_error ~extract_exn:true (fun () ->
+          let genesis_dir = dir ^/ Cache_dir.genesis_dir_name in
+          if Core.Sys.file_exists genesis_dir = `Yes then Deferred.return ()
+          else
+            (*Look for the tar and extract*)
+            let tar_file = genesis_dir ^ ".tar.gz" in
+            let%map _result =
+              Process.run_exn ~prog:"tar"
+                ~args:["-C"; dir; "-xzf"; tar_file]
+                ()
+            in
+            () )
+    with
+    | Ok () ->
+        ()
+    | Error e ->
+        Logger.debug ~module_:__MODULE__ ~location:__LOC__ logger
+          "Error extracting genesis ledger and proof : $error"
+          ~metadata:[("error", `String (Error.to_string_hum e))]
+  in
+  let retrieve dir =
+    Logger.debug ~module_:__MODULE__ ~location:__LOC__ logger
+      "Retrieving genesis ledger and genesis proof from $path"
+      ~metadata:[("path", `String dir)] ;
+    let%bind () = extract dir in
+    let dir = dir ^/ Cache_dir.genesis_dir_name in
+    let ledger_dir = dir ^/ "ledger" in
+    let proof_file = dir ^/ "genesis_proof" in
+    if
+      Core.Sys.file_exists ledger_dir = `Yes
+      && Core.Sys.file_exists proof_file = `Yes
+    then (
+      let genesis_ledger =
+        lazy (Ledger.create ~directory_name:ledger_dir ())
+      in
+      let%map base_proof =
+        match%map
+          Monitor.try_with_or_error ~extract_exn:true (fun () ->
+              let%bind r = Reader.open_file proof_file in
+              let%map contents =
+                Pipe.to_list (Reader.lines r) >>| String.concat
+              in
+              Sexp.of_string contents |> Proof.Stable.V1.t_of_sexp )
+        with
+        | Ok base_proof ->
+            base_proof
+        | Error e ->
+            failwithf "Error reading the base proof from %s: %s" proof_file
```

Comment:
We should raise a new exception here and handle it here https://github.com/CodaProtocol/coda/blob/develop/src/app/cli/src/init/coda_run.ml#L486-L497 so it doesn't look as bad to users.

Reply:
Is it ok if I make another PR for this? Requires some code rearranging